manpages/man/tiledb_generate_data.1
.TH tiledb_generate_data 1 "28 June 2015" "Version 0.1" "TileDB programs"
 
.SH NAME
tiledb_generate_data - generates synthetic data for a particular array schema
in a \fBCSV\fR or \fBbinary\fR form

.SH SYNOPSIS
.B tiledb_generate_data 
.BI "-w " "directory " "-A " "array_name " "-f " "filename " \
"-t " "filetype " "-n " "cell_num " 
[
.BI "-s " "seed "
]

.SH DESCRIPTION
.B tiledb_generate_data
generates sythetic data for a particular array schema
in a \fBCSV\fR or \fBbinary\fR form. On error, it prints a message on stderr and
returns an error code (see \fBEXIT STATUS\fR). 

The following conventions apply to the \fBSYNOPSIS\fR section:

.TS
tab (@);
c lx .
\fBbold text\fR @ type exactly as shown
\fIitalic text\fR @ replace with appropriate argument
[\fB\-a \fIarg\fR]@ any or all options within [ ] are optional
.TE

The dataset essentially consists of a collection of cells, whose format 
complies with the schema of an array given as argument to option \fB\-A\fR.
This means that the given array must already be defined. The number of cells to
be generated, as well as the seed to the random generator, are given as 
arguments to the program as well.

.TP
.B Data distribution
.I Cooridinates:
The coordinate values are drawn uniformly at random from their 
corresponding dimension domains defined in the array schema.

.I Attribute values:
The attribute values depend on their type. The real data values (i.e.,
\fBfloat\fR and \fBdouble\fR) are drawn uniformly at random from 
\fB[0.0,1.0]\fR. The \fBint\fR and \fBint64_t\fR data values are drawn uniformly
 at random from \fB[0,max()]\fR, where \fBmax()\fR is the maximum value of the
corresponding type supported by the system that calls the function. 
Finally, the \fBchar\fR values are drawn uniformly at random from ASCII
domain \fB[45,126]\fR, i.e., all the characters between and including \fB'-'\fR 
and \fB'~'\fR.

.TP
.B CSV file format
This is described in detail in \fBtiledb_load_csv(1)\fR. The difference is
that the generated data do not contain null values and deletions.

.TP
.B Binary file format
This is described in detail in \fBtiledb_load_bin(1)\fR. The difference is
that the generated data do not contain null values and deletions.

.SH OPTIONS
.TP
.BI "-w" " directory" "\fR, " \fB --workspace=\fIdirectory\fR  
The path to the workspace folder where the TileDB array is stored. The path
must exist, and the caller must have read and write permissions on it.

.TP
.BI "-A" " array_name" "\fR, " \fB --array-name=\fIarray_name\fR  
The name of the array for which the data are generated exported. The array
must be already defined. 

.TP
.BI "-f" " filename" "\fR, " \fB --filename=\fIfilename\fR  
The name of the output file that will store the generated data. 

.TP
.BI "-t" " filetype" "\fR, " \fB --filetype=\fIfiletype\fR  
The type of the generated file. It can be either \fBcsv\fR (CSV file) or
\fBbin\fR (binary file). 

.TP
.BI "-n" " cell_num" "\fR, " \fB --cell-num=\fIcell_num\fR  
The number of cells to be generated. 

.TP
.BI "-s" " seed" "\fR, " \fB --seed=\fIseed\fR  
The seed that will be used internally for the random generator. This is
optional. If it is omitted, the default is the current time of day. 

.SH EXAMPLES
.TP
\fBtiledb_generate_data\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-A \fRmy_array \\
    \fB-f \fRmy_array.csv \\
    \fB-t \fRcsv \\
    \fB-n \fR10000 \\
    \fB-s \fR 0

Creates CSV data for array \fImy_array\fR. The data consist of 10000 cells. A fixed 
seed \fB0\fR was used, e.g., for repeatability purposes.

\fBtiledb_generate_data\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-A \fRmy_array \\
    \fB-f \fRmy_array.bin \\
    \fB-t \fRbin \\
    \fB-n \fR10000

Creates binary data for array \fImy_array\fR. The data consist of 10000 cells. The 
default current time of day is used as a seed (which means that a different
dataset will be generated next time the program is executed with the same
options).

.SH EXIT STATUS
.TP 
.B 0
Successful execution
.TP 
.B -1
Parser error
.TP 
.B -2
Undefined array
.TP 
.B -3
File operation failed
.TP 
.B -7
TileDB initialization failed
.TP 
.B -8
TileDB finalization failed
.TP 
.B -14
Failed to open array
.TP 
.B -15
Failed to close array

.SH SEE ALSO
.BR "tiledb_define_array" "(1) " "tiledb_load_csv" "(1) " \
"tiledb_load_bin" "(1) "

.SH AUTHOR
Stavros Papadopoulos <http://people.csail.mit.edu/stavrosp>, Copyright (c) 2015


tiledb_cmd/src/tiledb_generate_data.cc
/**
 * @file   tiledb_generate_data.cc
 * @author Stavros Papadopoulos <stavrosp@csail.mit.edu>
 *
 * @section LICENSE
 *
 * The MIT License
 * 
 * @copyright Copyright (c) 2014 Stavros Papadopoulos <stavrosp@csail.mit.edu>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * 
 * @section DESCRIPTION
 *
 * Implements command "tiledb_generate_data".
 */

#include "csv_line.h"
#include "special_values.h"
#include "tiledb.h"
#include "utils.h"
#include <chrono>
#include <getopt.h>
#include <iostream>
#include <string>

/** 
 * Parses the command options. It returns 0 on success. On error, it prints
 * a message on stderr and returns -1.
 */                                                            
int parse_options(
    int argc, 
    char** argv, 
    std::string& array_name, 
    std::string& workspace,
    std::string& filename,
    std::string& filetype,
    unsigned int& seed,
    int64_t& cell_num) {
  // Initialization
  array_name = "";
  workspace = "";
  filename = "";
  filetype = "";
  std::string seed_str("");
  std::string cell_num_str("");

  // Define long options
  static struct option long_options[] = 
  {
    {"array-name",1,0,'A'},
    {"filename",1,0,'f'},
    {"seed",1,0,'s'},
    {"filetype",1,0,'t'},
    {"cell-num",1,0,'n'},
    {"workspace",1,0,'w'},
    {0,0,0,0},
  };
  // Define short options
  const char* short_options = "A:f:s:t:n:w:";
  // Get options
  int c;
  int option_num = 0;
  while((c=getopt_long(argc, argv, short_options, long_options, NULL)) >= 0) {
    ++option_num;
    switch(c) {
      case 'A':
        if(array_name != "") {
          std::cerr << ERROR_MSG_HEADER 
                    << " More than one array names provided.\n";
          return -1;
        }
        array_name = optarg;
        break;
      case 'f':
        if(filename != "") {
          std::cerr << ERROR_MSG_HEADER 
                    << " More than one file names provided.\n";
          return -1;
        }
        filename = optarg;
        break;
      case 's':
        if(seed_str != "") {
          std::cerr << ERROR_MSG_HEADER 
                    << " More than seeds provided.\n";
          return -1;
        }
        seed_str = optarg;
        break;
      case 't':
        if(filetype != "") {
          std::cerr << ERROR_MSG_HEADER 
                    << " More than one file types provided.\n";
          return -1;
        }
        filetype = optarg;
        break;
      case 'n':
        if(cell_num_str != "") {
          std::cerr << ERROR_MSG_HEADER 
                    << " More than one numbers of cells provided.\n";
          return -1;
        }
        cell_num_str = optarg;
        break;
      case 'w':
        if(workspace != "") {
          std::cerr << ERROR_MSG_HEADER 
                    << " More than one workspaces provided.\n";
          return -1;
        }
        workspace = optarg;
        break; 
      default:
        return -1;
    }
  }

  // Check number of arguments
  if((argc-1) != 2*option_num) {
    std::cerr << ERROR_MSG_HEADER << " Arguments-options mismatch.\n";
    return -1;
  }
  // Check if correct arguments have been ginen
  if(array_name == "") {
    std::cerr << ERROR_MSG_HEADER << " Array name not provided.\n";
    return -1;
  }
  if(workspace == "") {
    std::cerr << ERROR_MSG_HEADER << " Workspace not provided.\n";
    return -1;
  }
  if(filename == "") {
    std::cerr << ERROR_MSG_HEADER << " File name not provided.\n";
    return -1;
  }
  if(filetype == "") {
    std::cerr << ERROR_MSG_HEADER << " File type not provided.\n";
    return -1;
  }
  if(cell_num_str == "") {
    std::cerr << ERROR_MSG_HEADER << " Number of cells not provided.\n";
    return -1;
  }

  // Check number of arguments
  // ----- array_name
  CSVLine temp;
  temp << array_name;
  if(temp.val_num() > 1) {
    std::cerr << ERROR_MSG_HEADER 
              << " More than one array names provided.\n";
    return -1;
  }
  // ----- workspace
  temp.clear();
  temp << workspace;
  if(temp.val_num() > 1) {
    std::cerr << ERROR_MSG_HEADER 
              << " More than one workspaces provided.\n";
    return -1;
  }
  // ----- filename
  temp.clear();
  temp << filename;
  if(temp.val_num() > 1) {
    std::cerr << ERROR_MSG_HEADER 
              << " More than one files names provided.\n";
    return -1;
  }
  // ----- filetype
  temp.clear();
  temp << filetype;
  if(temp.val_num() > 1) {
    std::cerr << ERROR_MSG_HEADER 
              << " More than one file types provided.\n";
    return -1;
  }
  if(filetype != "csv" && filetype != "bin") {
    std::cerr << ERROR_MSG_HEADER 
              << " Unknown file type '" << filetype << "'.\n";
    return -1;
  }
  // ----- cell_num_str
  temp.clear();
  temp << cell_num_str;
  if(temp.val_num() > 1) {
    std::cerr << ERROR_MSG_HEADER 
              << " More than one numbers of cells provided.\n";
    return -1;
  }
  // ----- seed_str
  temp.clear();
  temp << seed_str;
  if(temp.val_num() > 1) {
    std::cerr << ERROR_MSG_HEADER 
              << " More than one seeds provided.\n";
    return -1;
  }

  // Get cell_num
  if(!is_non_negative_integer(cell_num_str.c_str())) {
    std::cerr << ERROR_MSG_HEADER 
              << " The number of cells must be a non-negative integer.\n";
    return -1;
  } else {
    sscanf(cell_num_str.c_str(), "%lld", &cell_num); 
  }

  // Get seed  
  if(seed_str == "") {
    seed = std::chrono::system_clock::now().time_since_epoch().count();
  } else if(!is_non_negative_integer(seed_str.c_str())) {
    std::cerr << ERROR_MSG_HEADER 
              << " The seed must be a non-negative integer.\n";
    return -1;
  } else {
    sscanf(seed_str.c_str(), "%u", &seed); 
  }
 
  return 0;
}

int main(int argc, char** argv) {
  // Arguments
  std::string array_name;
  std::string workspace;
  std::string filename;
  std::string filetype;
  unsigned int seed;
  int64_t cell_num;

  // Stores the return code of the various functions below
  int rc; 
 
  // Parse command line
  rc = parse_options(
           argc, argv, array_name, workspace, filename,
           filetype, seed, cell_num);
  if(rc) {
    std::cerr << ERROR_MSG_HEADER << " Failed to parse the command line.\n";
    return TILEDB_EPARSE;
  }

  // Initialize TileDB
  TileDB_CTX* tiledb_ctx;
  rc = tiledb_ctx_init(workspace.c_str(), tiledb_ctx);
  if(rc) {
    std::cerr << ERROR_MSG_HEADER << " Failed to initialize TileDB.\n";
    return TILEDB_EINIT;
  }

  // Generate data
  rc = tiledb_generate_data(
           tiledb_ctx, array_name.c_str(), filename.c_str(),
           filetype.c_str(), seed, cell_num); 
  if(rc) 
    return rc;

  // Finalize TileDB
  rc = tiledb_ctx_finalize(tiledb_ctx);
  if(rc) {
    std::cerr << ERROR_MSG_HEADER << " Failed to finalize TileDB.\n";
    return TILEDB_EFIN;
  }

  std::cout << MSG_HEADER << " Program executed successfully!\n";

  return 0;
}
